### PR TITLE
Support for Soft Prompts in PPO Model

### PR DESCRIPTION
**Pull request by daia99**
For adding support to soft prompts in models, fixes https://github.com/CarperAI/trlx/issues/29.

Changes:
- Sample config and example added with minor changes from original ppo versions.
- ModelConfig has two new params to support soft prompt settings.
- New accelerate model added with soft prompt implementation.

Not tested with full training run, but runs training steps in added example script.

### CODE REVIEW

File: trlx/data/configs.py [hunk 1]
```diff
@@ -29,6 +29,8 @@ class ModelConfig:
     model_type : str # One of the architectures present in framework.model
     device : str = ''
     num_layers_unfrozen : int = -1
+    n_soft_tokens: int = None # soft prompt support
```
Comment (LouisCastricato) [inline trlx/data/configs.py:32]: You should be making a new model config for this so it isnt inherited by other model configs except your own.

File: trlx/model/accelerate_ppo_softprompt_model.py [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+    def __init__(self, config, train_mode=True):
+        self.store = PPORolloutStorage()
+        super().__init__(config, self.store)
+
+        assert config.model.n_soft_tokens > 0, "Number of soft prompt tokens should be >=1"
+        
+        self.soft_dummy_token_id = 50256 # dummy token for padding soft prompts
+        self.n_soft_tokens = config.model.n_soft_tokens
+
+        s_wte = SoftEmbedding(self.model.gpt.get_input_embeddings(),
```
Comment (LouisCastricato) [inline trlx/model/accelerate_ppo_softprompt_model.py:88]: A few comments about what this does would be great.

File: examples/ppo_softprompt_sentiments.py [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+        sent_kwargs = {
+            "return_all_scores": True,
+            "function_to_apply": None,
+            "batch_size": cfg.method.chunk_size,
+        }
+        pipe_outputs = sentiment_pipe(samples, **sent_kwargs)
+        scores = torch.tensor([output[1]["score"] for output in pipe_outputs])
+        return scores
+
+    model: AcceleratePPOSoftpromptModel = get_model(cfg.model.model_type)(cfg)
```
Comment (LouisCastricato) [inline examples/ppo_softprompt_sentiments.py:33]: You don't need this. Just remove the type case from ppo_sentiments and leave the imports. The registery (e.g. get_model) will be able to determine which model type to load based off the config you use.

**Comment by LouisCastricato** [comment]
Needs work.

**Comment by Dahoas** [comment]
Thanks for the pr! It just occurred to me that I'm not sure how training soft-prompts will play with the usual practice of freezing the bottom 2/3 transformer blocks of the model. Surely there's some kind of work-around but I haven't thought of anything at the moment. 

At any rate things should be fine for small models which are fine-tuned without freezing any layers.

**Pull request closed without merging**